Support linux_gcc_arm64 arch in 6.7.0 via new linux_arm64 host

rectalogic opened the pull request:
Alternative implementation to #765 
Fixes #750

ddalcino commented:
This is probably a better solution than #765. I'm sure it will be easier to maintain.

However, a few things will need to be addressed:
* Update tests: Currently, none of our tests are doing anything to test that the `linux_arm64` host works on any of the `aqt list-*` or `install-*` commands. It would be bad to merge this without adding such tests.
* The `ci/generate_combinations.py` script doesn't know anything about `linux_arm64`, and it won't be able to make automatic changes to `aqt/combinations.json` on its own. Personally, I think it makes more sense to accept your `aqt/combinations.json` as-is than update the script, and just wait for resolution of #748, which should make `combinations.json` and `generate_combinations.py` unnecessary in the first place.
* I'd really like to test this with the Azure pipelines, to make sure it didn't break anything. However, the pipelines appear to be broken right now and since #735 I'm not sure how to trigger them to run.
* I'd also like to see an automated end-to-end test, like the existing tests in the Azure pipeline, that installs Qt for arm and compiles a simple project. Whether we want to do that in GH Actions or an Azure pipeline is an open question.

rectalogic commented:
>     * Update tests: Currently, none of our tests are doing anything to test that the `linux_arm64` host works on any of the `aqt list-*` or `install-*` commands. It would be bad to merge this without adding such tests.

Added some tests (and some fixes for bugs the tests uncovered).

>     * I'd really like to test this with the Azure pipelines, to make sure it didn't break anything. However, the pipelines appear to be broken right now and since [chore: update CI execution trigger/schedule #735](https://github.com/miurahr/aqtinstall/pull/735) I'm not sure how to trigger them to run.

I added Azure pipeline support, using a docker image running emulated arm64. But this is all untested since I don't know how to trigger it either...

miurahr commented on ci/generate_azure_pipelines_matrices.py:
```diff
@@ -92,11 +92,13 @@ def __init__(self, platform, build_jobs):
 
 all_platform_build_jobs = [
     PlatformBuildJobs("linux", linux_build_jobs),
+    PlatformBuildJobs("linux_arm64", linux_arm64_build_jobs),
```
I don't think it is valuable  test it for all versions of targets.

rectalogic replied:
I'm not clear how you want to limit it - should `PlatformBuildJobs` take a subset of python versions or something? It looks like `PlatformBuildJobs` is how we get variables into matrices which is used in `azure-pipelines.yml` to actually run tests via `steps.yml`? I'm not really familiar with azure pipelines so I was just following how the other platforms work, and linux_arm64 is a new platform.

miurahr replied:
Sorry for my misunderstanding.  My comment here is incorrect.

> I'm not clear how you want to limit it - should `PlatformBuildJobs` take a subset of python versions or something? It looks like `PlatformBuildJobs` is how we get variables into matrices which is used in `azure-pipelines.yml` to actually run tests via `steps.yml`? I'm not really familiar with azure pipelines so I was just following how the other platforms work, and linux_arm64 is a new platform.

A code chunk
```python
print("Setting Variables below")
print(
    f"##vso[task.setVariable variable=linux;isOutput=true]{json.dumps(matrices['linux'])}"
)
```
defines output matrix variable "linux"  that passed to `azure-pipelines.yml` 
then Linux tasks are defined as
```yaml
- job: Linux
  dependsOn: MatricesGenerator
  pool:
    vmImage: 'ubuntu-22.04'
  strategy:
    matrix: $[ dependencies.MatricesGenerator.outputs['mtrx.linux'] ]
  steps:
    - template: ci/steps.yml
```

Here it refers MatirxGenerator output `linux` variable.
The actual steps are defined in `ci/steps.yml`

 `PlatformBuildJobs("linux_arm64"` defines a definitions of platform "linux_arm64" and output as json.
A defined jobs will be referenced like

```yaml
  strategy:
    matrix: $[ dependencies.MatricesGenerator.outputs['mtrx.linux_arm64'] ]
```
 in `azure-pipelines.yml`.

rectalogic replied:
> Sorry for my misunderstanding. My comment here is incorrect.

I'm not clear if you are still asking for changes or not.

> A code chunk
> 
> ```python
> print("Setting Variables below")
> print(
>     f"##vso[task.setVariable variable=linux;isOutput=true]{json.dumps(matrices['linux'])}"
> )
> ```
> 
> defines output matrix variable "linux" that passed to `azure-pipelines.yml` 

I am doing this for linux_arm64 here https://github.com/rectalogic/aqtinstall/blob/b51d200c32fd948f6e918c5ecada1dc621e68161/ci/generate_azure_pipelines_matrices.py#L486-L488

> then Linux tasks are defined as
> 
> ```yaml
> - job: Linux
>   dependsOn: MatricesGenerator
>   pool:
>     vmImage: 'ubuntu-22.04'
>   strategy:
>     matrix: $[ dependencies.MatricesGenerator.outputs['mtrx.linux'] ]
>   steps:
>     - template: ci/steps.yml
> ```

I am doing this for linux_arm64 here https://github.com/rectalogic/aqtinstall/blob/b51d200c32fd948f6e918c5ecada1dc621e68161/azure-pipelines.yml#L57-L67
This is where I use a docker container on azure to run arm64 platform (not sure this will work, needs testing)


> `PlatformBuildJobs("linux_arm64"` defines a definitions of platform "linux_arm64" and output as json. A defined jobs will be referenced like
> 
> ```yaml
>   strategy:
>     matrix: $[ dependencies.MatricesGenerator.outputs['mtrx.linux_arm64'] ]
> ```
> 
> in `azure-pipelines.yml`.

https://github.com/rectalogic/aqtinstall/blob/b51d200c32fd948f6e918c5ecada1dc621e68161/azure-pipelines.yml#L65

miurahr replied:
I see. Now I can understand and LGTM.

miurahr commented on ci/generate_azure_pipelines_matrices.py:
```diff
@@ -368,32 +371,36 @@ def __init__(self, platform, build_jobs):
-)
+for lst, os_name, version, arch, arch_dir in (
+    (linux_build_jobs, "linux", "5.12.11", "gcc_64", "gcc_64"),
+    (linux_arm64_build_jobs, "linux_arm64", "6.7.0", "linux_gcc_arm64", "gcc_arm64"),
```
Why you test linux_arm64 with long path test case?

rectalogic replied:
I think this test is patching the `qmake` binary and so I thought there may be differences in patching an amd64 binary vs arm64. I can remove it if you think it's redundant.

ddalcino replied:
@rectalogic please see #257, #258, and #259 for context regarding these tests. IMHO, the existing tests are already sufficient for what they’re intended to test, and I don’t see a reason to add anything more here. I appreciate that you’re trying to be thorough though.

The reality is that Azure compute time isn’t cheap, and we don’t want to waste it on tests that won’t tell us much new information. Otherwise, we would be testing this property on Mac and Windows as well.

ddalcino commented:
I forgot to mention, our documentation will need to be updated as well. So far, all the docs mention that the 'host' argument is only allowed to be Mac, Windows, or Linux; this won't be true anymore.

It's probably worthwhile to explain somewhere why Linux is now split between two different hosts: `linux` and `linux_arm64`, and why that isn't the case for Mac and Windows. As I understand it, after a certain version of Qt 6 (6.2 maybe?), all the Mac Qt binaries became 'universal' binaries that include both x64 and arm64. I could be wrong about that; it's been a while since I looked.

I'm not sure if there are any remaining places in the docs, or CLI help text messages, that refer to 'host' and 'operating system' interchangeably, but if there are, that should probably be cleaned up. Do `linux` and `linux_arm64` count as different operating systems? IMHO, 'host' expresses the tuple [OS, processor architecture], and that's distinct from our 'architecture' parameter, which expresses the tuple [compiler, processor architecture]. It's a bit messy, and deserves a clear explanation of what all these terms mean.

rectalogic commented:
6.7 is released
https://www.qt.io/blog/qt-6.7-released

Should we merge this and do a new aqtinstall release?

rectalogic commented:
@miurahr can you merge this? Currently even amd64 linux 6.7.0 is not installable with aqt 3.1.12 even if you use the new arch name (`linux_gcc_64` instead of `gcc_64`):

```sh-session
> aqt install-qt linux desktop 6.7.0 -O /tmp/qt
INFO    : aqtinstall(aqt) v3.1.12 on Python 3.11.8 [CPython Clang 15.0.0 (clang-1500.1.0.2.5)]
ERROR   : The packages ['qt_base'] were not found while parsing XML of package information!
==============================Suggested follow-up:==============================
* Please use 'aqt list-qt linux desktop --arch 6.7.0' to show architectures available.

> aqt install-qt linux desktop 6.7.0 linux_gcc_64 -O /tmp/qt
INFO    : aqtinstall(aqt) v3.1.12 on Python 3.11.8 [CPython Clang 15.0.0 (clang-1500.1.0.2.5)]
INFO    : Downloading qtsvg...
INFO    : Downloading qttools...
INFO    : Downloading qtdeclarative...
INFO    : Downloading qtbase...
INFO    : Redirected: qt.mirror.constant.com
INFO    : Redirected: qt.mirror.constant.com
INFO    : Redirected: qt.mirror.constant.com
INFO    : Redirected: qt.mirror.constant.com
INFO    : Finished installation of qtsvg-Linux-RHEL_8_8-GCC-Linux-RHEL_8_8-X86_64.7z in 1.18460154
INFO    : Downloading qttranslations...
INFO    : Redirected: qt.mirror.constant.com
INFO    : Finished installation of qttranslations-Linux-RHEL_8_8-GCC-Linux-RHEL_8_8-X86_64.7z in 1.32922746
INFO    : Downloading qtwayland...
INFO    : Redirected: qt.mirror.constant.com
INFO    : Finished installation of qtwayland-Linux-RHEL_8_8-GCC-Linux-RHEL_8_8-X86_64.7z in 1.12197392
INFO    : Downloading icu...
INFO    : Finished installation of qtbase-Linux-RHEL_8_8-GCC-Linux-RHEL_8_8-X86_64.7z in 4.41483796
INFO    : Redirected: qt.mirror.constant.com
INFO    : Finished installation of qttools-Linux-RHEL_8_8-GCC-Linux-RHEL_8_8-X86_64.7z in 5.18647496
INFO    : Finished installation of icu-linux-Rhel8.6-x86_64.7z in 2.62416004
INFO    : Finished installation of qtdeclarative-Linux-RHEL_8_8-GCC-Linux-RHEL_8_8-X86_64.7z in 11.03515671
ERROR   : Updater caused an IO error: [Errno 2] No such file or directory: '/tmp/qt/6.7.0/linux_gcc_64/mkspecs/qconfig.pri'
```
